airbyte-cdk/java/airbyte-cdk/core/src/main/kotlin/io/airbyte/cdk/core/IntegrationCommand.kt

```diff
@@ -0,0 +1,165 @@
+            ],
+    )
+    lateinit var catalogFile: Path
+
+    @CommandLine.Option(
+        names = ["--" + JavaBaseConstants.ARGS_STATE_KEY],
+        description =
+            [
+                JavaBaseConstants.ARGS_PATH_DESC,
+                "Required by the following commands: read",
+            ],
+    )
+    lateinit var stateFile: Optional<Path>
+
+    @CommandLine.Spec lateinit var commandSpec: CommandLine.Model.CommandSpec
+
+    override fun run() {
+        val result = execute(command)
+        result.onSuccess { airbyteMessage ->
+            airbyteMessage?.let { outputRecordCollector.accept(airbyteMessage) }
+        }
+        result.onFailure {
+            commandSpec.commandLine().usage(System.out)
+            logger.error(it) { "\nUnable to perform operation '$command'." }
+            // Many of the exceptions thrown are nested inside layers of RuntimeExceptions. An
+            // attempt is made
+            // to
+            // find the root exception that corresponds to a configuration error. If that does not
+            // exist, we
+            // just return the original exception.
+            ApmTraceUtils.addExceptionToTrace(it)
+            val rootThrowable = ConnectorExceptionUtil.getRootConfigError(Exception(it))
+            val displayMessage = ConnectorExceptionUtil.getDisplayMessage(rootThrowable)
+            // If the source connector throws a config error, a trace message with the relevant
+            // message should
+            // be surfaced.
+            if (ConnectorExceptionUtil.isConfigError(rootThrowable)) {
+                AirbyteTraceMessageUtility.emitConfigErrorTrace(it, displayMessage)
+            }
+            if (OperationType.CHECK.name.equals(command, true)) {
+                // Currently, special handling is required for the CHECK case since the user display
+                // information in
+                // the trace message is
+                // not properly surfaced to the FE. In the future, we can remove this and just throw
+                // an exception.
+                outputRecordCollector.accept(
+                    AirbyteMessage()
+                        .withType(AirbyteMessage.Type.CONNECTION_STATUS)
+                        .withConnectionStatus(
+                            AirbyteConnectionStatus()
+                                .withStatus(AirbyteConnectionStatus.Status.FAILED)
+                                .withMessage(displayMessage),
+                        ),
+                )
+            }
+        }
+
+        logger.info { "Completed integration: $connectorName" }
+    }
+
+    private fun execute(command: String): Result<AirbyteMessage?> {
+        return try {
+            val operationType = OperationType.valueOf(command.uppercase())
+            val operation: Operation? =
+                operations.firstOrNull { o: Operation -> (operationType == o.type()) }
```

Comment:
Can you just throw in case there's not exactly one operation? Errors are going to be weird otherwise.

Reply:
@postamar Since we are using the Kotlin `Result` object here, we don't want to throw just to catch and return the result.  Therefore, this is checking to see if the operation is there and if not, returning in the failure result.  This could be re-written to use an if expression in Kotlin if that would be easier to read/understand:

Reply:
I meant more along the lines of let's fail this thing when there's more than one operation. There should only be exactly one, right? 

Personally I don't mind the `?` monad at all.

Reply:
@postamar I see the confusion -- theoretically yes, there should be only one that matches the type.  However, the list of operations injected into the class may contain more than one, as it is based on the conditional logic put on the class for Micronaut.  There is nothing that will enforce that people do the correct thing and limit the creation of the operation bean to when the operation is selected.  This is how I have implemented it in the CDK, but since each connector can provide replacements for these beans, we can't guarantee that.  If we want to say that the CDK will enforce that and fail loudly if they do that, I can get on board with that, but I want to be very clear that this is something that the CDK enforces.

Reply:
I've created a ticket to fix this:  https://github.com/airbytehq/airbyte-internal-issues/issues/6723

Reply:
I don't understand why we can't have a for-loop that loops through all of the operations and checks that there's a 1-1 mapping between operation type and enum values and throws an exception otherwise.

> If we want to say that the CDK will enforce that and fail loudly if they do that, I can get on board with that, but I want to be very clear that this is something that the CDK enforces.

Basically, yes. What's the big deal? Am I missing something? (I might)

Reply:
+1 to what Marius said

Reply:
@postamar @stephane-airbyte We can definitely do that.  My point is that is not a technical limitation to doing that -- it's a product one from a CDK point of view.  If we agree that the CDK should blow up the connector if they don't configure the conditional Singleton creation to only create one operation that matches the provided command, then we need to advertise that as part of the CDK and enforce it.  I'm not the person to make that decision and I'm not sure at this point that we won't run into some case where we need this to be relaxed as we go, so my default here is to leave it flexible for now and restrict it later once we have converted a few connectors and have a better sense about this.

Reply:
Understood. I don't think anyone will come forward claiming ownership on these matters so here, I'll make that decision: let's blow up the connector when there's more than one command.  I'll take take responsibility for it and its consequences.

Reply:
@postamar We can solve this simply by changing the injection from a list to a single command.  Micronaut will handle exploding if it detects more than one candidate for the injection.  If that sounds reasonable, I can make the change now in this PR.

Reply:
In case it wasn't obvious yet: yes please.